File: Problems/0203_Remove_Linked_List_Elements/Remove_Linked_List_Elements.py
# coding: utf-8

import os
import sys
import time

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def removeElements(self, head, val):
        dummy = ListNode(-1)
        curr = dummy
        while head:
            if head.val != val:
                curr.next = head
                curr = curr.next
                head = head.next
            else:
                head = head.next
        curr.next = None
        return dummy.next


    def removeElements_work(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        temp_node = head
        while temp_node != None:
            if temp_node.next != None:
                if temp_node.next.val == val:
                    temp_node.next = temp_node.next.next
            temp_node = temp_node.next
        return head


def set_node(data):
    if len(data) == 0:
        return None
    node = ListNode(data[0])
    temp_node = node

    for i in range(1,len(data)):
        temp_node.next = ListNode(data[i])
        temp_node = temp_node.next
    
    return node


def output_node(node):
    if node == None:
        return ''
    tempStr = str(node.val)
    temp_node = node.next
    while temp_node != None:
        tempStr += ',' + str(temp_node.val)
        temp_node = temp_node.next
    return tempStr


def array_str_to_int(numbersStr):
    if len(numbersStr) == 0:
        return ""
    if numbersStr[0] == '':
        return ""
    numbers = [0]*len(numbersStr)
    for i in range(len(numbersStr)):
        numbers[i] = int(numbersStr[i])
    return numbers


def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)


def loop_main(temp):
    flds = temp.split(", ")
    numsStr = flds[0].replace("[","").replace("]","").split(",")
    nums = array_str_to_int(numsStr)
    val = int(flds[1])

    head = set_node(nums)
    print("nodes = %s" %(output_node(head)))

    time0 = time.time()

    sl = Solution()
    result = sl.removeElements(head, val)
    print("result = %s" %(output_node(result)))

    time1 = time.time()

    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0203_Remove_Linked_List_Elements/Remove_Linked_List_Elements.cs
using System;

// Definition for singly-linked list.

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int x) {
        val = x;
        next = null;
    }
}

public class Solution {
    public ListNode RemoveElements(ListNode head, int val)
    {
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;

        while (head != null) {
            if (head.val != val) {
                curr.next = head;
                curr = curr.next;
                head = head.next;
            }
            else
                head = head.next;
        }

        curr.next = null;
        return dummy.next;
    }
    
    private ListNode set_node(string[] data)
    {
        ListNode node, temp_node;
        
        if (data.Length <= 0)
            return null;

        if (data[0] == "")
            return null;

        node = new ListNode(int.Parse(data[0]));
        temp_node = node;

        for (int i = 1; i < data.Length; i++) {
            temp_node.next = new ListNode(int.Parse(data[i]));
            temp_node = temp_node.next;
        }
        
        return node;
    }
    
    private string output_node(ListNode node)
    {
        if (node == null)
            return "";

        ListNode temp_node = node;
        string resultStr = "";

        if (node == null) {
            return resultStr;
        }
        
        resultStr += temp_node.val.ToString();
        temp_node = temp_node.next;

        while (temp_node != null) {
            resultStr += "," + temp_node.val.ToString();
            temp_node = temp_node.next;
        }
        
        return resultStr;
    }

    public void Main(string args)
    {
        string[] flds = args.Split(new string[] {", "}, StringSplitOptions.None);
        string[] data = flds[0].Replace("[","").Replace("]","").Split(',');
        int k = int.Parse(flds[1]);

        ListNode node = set_node(data);
        Console.WriteLine("node = " + output_node(node));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        ListNode result_node = RemoveElements(node, k);
        Console.WriteLine("Result = " + output_node(result_node));
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}
